framework/src/main/java/org/checkerframework/framework/type/QualifierHierarchy.java

```diff
@@ -81,361 +100,351 @@ public int getWidth() {
-     *
-     * <p>Examples:
-     *
-     * <ul>
-     *   <li>For NonNull, leastUpperBound('Nullable', 'NonNull') &rarr; Nullable
-     * </ul>
-     *
-     * The two qualifiers have to be from the same qualifier hierarchy. Otherwise, null will be
-     * returned.
-     *
-     * <p>This method works even if the underlying Java type is a type variable. In that case, a
-     * 'null' AnnnotationMirror and the empty set represent a meaningful value (namely, no
-     * annotation).
+     * Returns the annotation in annos that is in the hierarchy for which annotationMirror is top.
      *
-     * @return the least restrictive qualifiers for both types
+     * @param annos set of annotations to search
+     * @param top the top annotation in the hierarchy to which the returned annotation belongs
+     * @return annotation in the same hierarchy as annotationMirror, or null if one is not found
      */
-    public abstract AnnotationMirror leastUpperBoundTypeVariable(
-            AnnotationMirror a1, AnnotationMirror a2);
+    default @Nullable AnnotationMirror findAnnotationInHierarchy(
+            Collection<? extends AnnotationMirror> annos, AnnotationMirror top) {
+        for (AnnotationMirror anno : annos) {
+            if (isSubtype(anno, top)) {
+                return anno;
+            }
+        }
+        return null;
+    }
 
     /**
-     * Returns the greatest lower bound for the qualifiers a1 and a2.
-     *
-     * <p>The two qualifiers have to be from the same qualifier hierarchy. Otherwise, null will be
-     * returned.
-     *
-     * <p>This method works even if the underlying Java type is a type variable. In that case, a
-     * 'null' AnnnotationMirror and the empty set represent a meaningful value (namely, no
-     * annotation).
+     * Update a mapping from {@code key} to a set of AnnotationMirrors. If {@code key} is not
+     * already in the map, then put it in the map with a value of a new set containing {@code
+     * qualifier}. If the map contains {@code key}, then add {@code qualifier} to the set to which
+     * {@code key} maps. If that set contains a qualifier in the same hierarchy as {@code
+     * qualifier}, then don't add it and return false.
      *
-     * @param a1 first annotation
-     * @param a2 second annotation
-     * @return greatest lower bound of the two annotations
+     * @param map the mapping to modify
+     * @param key the key to update
+     * @param qualifier the value to add
+     * @param <T> type of the map's keys
+     * @return true if the update was done; false if there was a qualifier hierarchy collision
      */
-    public abstract AnnotationMirror greatestLowerBoundTypeVariable(
-            AnnotationMirror a1, AnnotationMirror a2);
+    default <T> boolean updateMappingToMutableSet(
+            Map<T, Set<AnnotationMirror>> map, T key, AnnotationMirror qualifier) {
+        @SuppressWarnings("nullness:argument.type.incompatible") // key is of type T.
```

Comment:
I'm confused about the suppression. Both the variable `key` and the key-type of `map` are `T`, so just saying `key is of type T`, doesn't explain why there is an error that can be suppressed.

Reply:
The type of the parameter of `containsKey` is `@NonNull Object`.  Call is legal because key is of type `T`.  I added a reference to https://github.com/typetools/checker-framework/issues/2000.